latch_cli/services/cp/upload/worker.py

```diff
@@ -0,0 +1,206 @@
+    url: str,
+    index: int,
+    part_size: int,
+    pbar: tqdm.tqdm,
+) -> CompletedPart:
+    with open(src, "rb") as f:
+        data = os.pread(f.fileno(), part_size, index * part_size)
+
+    res = await session.put(url, data=data)
+    if res.status != 200:
+        raise RuntimeError(f"failed to upload part {index} of {src}: {res.content}")
+
+    etag = res.headers["ETag"]
+    assert etag is not None, (
+        f"Malformed response from chunk upload for {src}, Part {index},"
+        f" Headers: {res.headers}"
+    )
+
+    pbar.update(len(data))
+
+    return CompletedPart(src=src, etag=etag, part=index + 1)
+
+
+min_part_size = 5 * Units.MiB
+
+
+async def work_loop(
+    work_queue: asyncio.Queue[Work],
+    total_pbar: tqdm.tqdm,
+    show_task_progress: bool,
+    print_file_on_completion: bool,
+):
+    pbar = tqdm.tqdm(
+        total=0,
+        leave=False,
+        unit="B",
+        unit_scale=True,
+        disable=not show_task_progress,
+    )
+
+    async with RetryClientSession(read_timeout=90, conn_timeout=10) as sess:
+        while True:
+            try:
+                work = work_queue.get_nowait()
+            except asyncio.QueueEmpty:
+                break
+
+            resolved = work.src
+            if work.src.is_symlink():
+                resolved = work.src.resolve()
+
+            content_type, _ = mimetypes.guess_type(resolved)
+            if content_type is None:
+                with open(resolved, "rb") as f:
+                    sample = f.read(Units.KiB)
+
+                try:
+                    sample.decode()
+                    content_type = "text/plain"
+                except UnicodeDecodeError:
+                    content_type = "application/octet-stream"
+
+            file_size = resolved.stat().st_size
+            if file_size > latch_constants.maximum_upload_size:
+                raise ValueError(
+                    f"{resolved}: file is {with_si_suffix(file_size)} which exceeds the"
+                    " maximum upload size (5TiB)",
+                )
+
+            chunk_size = work.chunk_size_mib * Units.MiB
+            if chunk_size < min_part_size:
+                click.secho(
+                    """\
+                    Unable to complete upload - please check your internet connection speed or any firewall settings that may block outbound traffic.\
+                    """,
+                    fg="red",
+                )
+                raise click.exceptions.Exit(1)
+
+            part_count = min(
+                latch_constants.maximum_upload_parts,
+                math.ceil(file_size / chunk_size),
+            )
+            part_size = max(
+                chunk_size,
+                math.ceil(file_size / latch_constants.maximum_upload_parts),
+            )
+
+            pbar.desc = resolved.name
+            pbar.total = file_size
+
+            # jitter to not dos nuc-data
+            await asyncio.sleep(0.1 * random.random())
```

Comment:
kinda odd what did we see here before?

Reply:
for wide directories with small files theres not enough time between `start-upload` calls so we end up throttling nuc-data

Reply:
But like shouldn't we use a semaphore on the call rather than adding jitter? Jitter is worse because it is not aware of how many calls are inflight or how long they are taking.